[BUG] Not Retryable Exceptions are ignored when executing tasks with `runStepOnce`

### JobRunr Version

8

### JDK Version

NA

### Your SQL / NoSQL database

NA

### What happened?

I'm using JobRunr's `doNotRetry` flag on `JobRunrException` with `runStepOnce` but the exception is converted to a `StepExecutionException` and the flag is no longer taken into account.

More precisely, I throw the following extension of JobRunrException:

```java
public class NonRetryableException extends JobRunrException {

    public NonRetryableException(String message) {
        super(message, true);
    }
}
``` 

### How to reproduce?

```java
   void myJobWithSteps() {
        context.runStepOnce("transfer-money", () -> transferMoney(payment));
   }

   void transferMoney(Payment payment) {
        throw new NonRetryableException("Credit card not found");
   }
```


### Relevant log output

```shell
org.jobrunr.jobs.exceptions.StepExecutionException: Exception during execution of step 'transfer-money'
	at org.jobrunr.jobs.context.JobContext.runStepOnce(JobContext.java:224)
	at org.jobrunr.storylinedemo.payments.PaymentService.processPayment(PaymentService.java:90)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:565)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:359)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:715)
	at org.jobrunr.storylinedemo.payments.PaymentService$$SpringCGLIB$$0.processPayment(<generated>)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:565)
	at org.jobrunr.server.runner.AbstractBackgroundJobRunner$BackgroundJobWorker.invokeJobMethod(AbstractBackgroundJobRunner.java:86)
	at org.jobrunr.server.runner.AbstractBackgroundJobRunner$BackgroundJobWorker.run(AbstractBackgroundJobRunner.java:58)
	at org.jobrunr.server.runner.AbstractBackgroundJobRunner.run(AbstractBackgroundJobRunner.java:34)
	at org.jobrunr.server.BackgroundJobPerformer.runActualJob(BackgroundJobPerformer.java:102)
	at org.jobrunr.server.BackgroundJobPerformer.performJob(BackgroundJobPerformer.java:71)
	at org.jobrunr.server.BackgroundJobPerformer.run(BackgroundJobPerformer.java:48)
	at java.base/java.util.concurrent.ThreadPerTaskExecutor$TaskRunner.run(ThreadPerTaskExecutor.java:291)
	at java.base/java.lang.VirtualThread.run(VirtualThread.java:456)
Caused by: org.jobrunr.storylinedemo.exceptions.NonRetryableException: Could not find the target credit card 5314586970914451
```

--- a/core/src/main/java/org/jobrunr/jobs/exceptions/StepExecutionException.java
+++ b/core/src/main/java/org/jobrunr/jobs/exceptions/StepExecutionException.java
@@ -1,11 +1,17 @@
 package org.jobrunr.jobs.exceptions;
 
+import org.jobrunr.JobRunrException;
+
 /**
  * An exception that is thrown if a certain step inside a Job fails
  */
-public class StepExecutionException extends RuntimeException {
+public class StepExecutionException extends JobRunrException {
 
     public StepExecutionException(String message, Exception exception) {
-        super(message, exception);
+        super(message, isJobRunrExceptionThatMustNotBeRetried(exception), exception);
+    }
+
+    private static boolean isJobRunrExceptionThatMustNotBeRetried(Exception exception) {
+        return exception instanceof JobRunrException && ((JobRunrException) exception).isProblematicAndDoNotRetry();
     }
 }

--- a/core/src/test/java/org/jobrunr/jobs/context/JobContextTest.java
+++ b/core/src/test/java/org/jobrunr/jobs/context/JobContextTest.java
@@ -1,6 +1,7 @@
 package org.jobrunr.jobs.context;
 
 import org.assertj.core.api.Assertions;
+import org.jobrunr.JobRunrException;
 import org.jobrunr.jobs.Job;
 import org.jobrunr.jobs.exceptions.StepExecutionException;
 import org.jobrunr.jobs.mappers.JobMapper;
@@ -197,6 +198,20 @@ public class JobContextTest {
     }
 
     @Test
+    void jobContextRunsStepOnlyOnceSupplierCanThrowNonRetryableException() {
+        final Job job = aJobInProgress().withName("job1").withLabels("my-label").build();
+
+        JobContext jobContext = new JobContext(job);
+
+        assertThatCode(() -> jobContext.runStepOnce("my-step", this::getSomethingThatThrowsAnNonRetryableException))
+                .isInstanceOf(StepExecutionException.class)
+                .isInstanceOf(JobRunrException.class)
+                .hasMessageContaining("Exception during execution of step 'my-step'")
+                .isInstanceOfSatisfying(JobRunrException.class, x -> assertThat(x.isProblematicAndDoNotRetry()).isTrue());
+    }
+
+
+    @Test
     void jobContextIsThreadSafeUsingJackson() throws InterruptedException {
         jobContextIsThreadsafe(new JobMapper(new JacksonJsonMapper()));
     }
@@ -286,5 +301,9 @@ public class JobContextTest {
             throw new Exception("Something went wrong");
         }
         return "test-" + counter.get();
+    }
+
+    private void getSomethingThatThrowsAnNonRetryableException() {
+        throw new JobRunrException("Something went wrong", true);
     }
 }

--- a/core/src/test/java/org/jobrunr/jobs/exceptions/StepExecutionExceptionTest.java
+++ b/core/src/test/java/org/jobrunr/jobs/exceptions/StepExecutionExceptionTest.java
@@ -1,0 +1,31 @@
+package org.jobrunr.jobs.exceptions;
+
+import org.jobrunr.JobRunrException;
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class StepExecutionExceptionTest {
+
+
+    @Test
+    void ifExceptionIsNormalExceptionThenDoNotRetryIsFalse() {
+        StepExecutionException e = new StepExecutionException("test", new Exception());
+
+        assertThat(e.isProblematicAndDoNotRetry()).isFalse();
+    }
+
+    @Test
+    void ifExceptionIsJobRunrExceptionWithIsProblematicAndDoNotRetrySetToFalseThenDoNotRetryIsFalse() {
+        StepExecutionException e = new StepExecutionException("test", new JobRunrException("an exception"));
+
+        assertThat(e.isProblematicAndDoNotRetry()).isFalse();
+    }
+
+    @Test
+    void ifExceptionIsJobRunrExceptionWithIsProblematicAndDoNotRetrySetToTrueThenDoNotRetryIsTrue() {
+        StepExecutionException e = new StepExecutionException("test", new JobRunrException("an exception", true));
+
+        assertThat(e.isProblematicAndDoNotRetry()).isTrue();
+    }
+}

